### PR TITLE
ext2: Add ability to mount devices and filesystems

### PR DESCRIPTION
**Pull request by jmaksymowicz**
Added the ability to mount a different filesystem within an ext2 filesystem.

## Description
Now another filesystem can be mounted using a directory within ext2 as the mountpoint. 

Additionally, the ext2 filesystem driver had a path length limit of 255 characters. This limitation has been removed and now paths of any length can be looked up within ext2. Single file names are still limited to 255 characters, as this is a file system limitation.

This also fixes issue https://github.com/phoenix-rtos/phoenix-rtos-project/issues/749 (on ext2 only).

## Motivation and Context
Fixes https://github.com/phoenix-rtos/phoenix-rtos-project/issues/114

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [x] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: ia32-generic-qemu, armv7a9-zynq7000-zturn

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Comment by nalajcie** [comment]
I believe it also fixes https://github.com/phoenix-rtos/phoenix-rtos-project/issues/114, please add appropriate comment in commit message.

File: ext2/ext2.c [hunk 1]
```diff
@@ -74,29 +75,32 @@ int ext2_destroy(ext2_t *fs, id_t id)
...
+	if ((len == 0) || (name == NULL)) {
 		return -EINVAL;
+	}
 
-	if ((dir = ext2_obj_get(fs, id)) == NULL)
+	if ((dir = ext2_obj_get(fs, id)) == NULL) {
 		return -ENOENT;
+	}
 
+	mutexLock(dir->lock);
```
Comment (agkaminski) [inline ext2/ext2.c:93]: Hmm, now we might take `dir->lock` once, but unlock it many times in for below, don't we?

File: ext2/ext2.c [hunk 2]
```diff
@@ -122,20 +129,18 @@ int ext2_lookup(ext2_t *fs, id_t id, const char *name, uint8_t len, oid_t *res,
...
-	mutexLock(obj->lock);
+		}
 
-	if (S_ISCHR(obj->inode->mode) || S_ISBLK(obj->inode->mode)) {
-		memcpy(dev, &obj->dev, sizeof(oid_t));
-	}
-	else {
-		dev->port = fs->port;
-		dev->id = res->id;
+		mutexLock(obj->lock);
```
Comment (agkaminski) [inline ext2/ext2.c:136]: lock taken in `for`, released only after exiting from `for` - might be taken many times (i.e. 2 times, resulting in deadlock)

**Review by agkaminski — COMMENTED** [review]
I think something strange has happened to locks in `ext2_lookup`. Perhaps `dir = obj` missing somewhere?

**Comment by jmaksymowicz** [comment]
> I think something strange has happened to locks in `ext2_lookup`. Perhaps `dir = obj` missing somewhere?

It is in the "increment" portion of the `for` statement.

File: ext2/ext2.c (hunk 1 above)
Comment (jmaksymowicz) [reply ext2/ext2.c:93]: This works because the object pointed to by `dir` changes with each iteration. I could rewrite this loop to make it more readable.

**Comment by agkaminski** [comment]
> > I think something strange has happened to locks in `ext2_lookup`. Perhaps `dir = obj` missing somewhere?
> 
> It is in the "increment" portion of the `for` statement.

Only place I haven't looked for it :)

**Review by agkaminski — APPROVED** [review]
LGTM

File: ext2/ext2.c (hunk 1 above)
Comment (agkaminski) [reply ext2/ext2.c:93]: It's ok, let's keep style/structure improvements as a separate task for later

**Pull request merged**